keps/sig-apps/961-maxunavailable-for-statefulset/README.md

```diff
@@ -32,56 +134,155 @@
+   time. If we used StatefulSet, we can only go one pod at a time which would result in a slow rolling
+   update. If StatefulSet supported maxUnavailable with value greater than 1, it would allow for a faster
+   rollout since a total of maxUnavailable number of pods could be loading up the cache at the same time.
+3. My Stateful clustered application, has followers and leaders, with followers being many more than 1. My application can tolerate many followers going down at the same time. I want to be able to do faster
+   rollouts by bringing down 2 or more followers at the same time. This is only possible if StatefulSet
+   supports maxUnavailable in Rolling Updates.
+4. Sometimes I just want easier tracking of revisions of a rolling update. Deployment does it through
+   ReplicaSets and has its own nuances. Understanding that requires diving into the complicacy of hashing
+   and how ReplicaSets are named. Over and above that, there are some issues with hash collisions which
+   further complicate the situation(I know they were solved). StatefulSet introduced ControllerRevisions
+   in 1.7 which are much easier to think and reason about. They are used by DaemonSet and StatefulSet for
+   tracking revisions. It would be so much nicer if all the use cases of Deployments can be met in
+   StatefulSet's and additionally we could track the revisions by ControllerRevisions. Another way of
+   saying this is, all my Deployment use cases are easily met by StatefulSet, and additionally I can enjoy
+   easier revision tracking only if StatefulSet supported `maxUnavailable`.
+
+With this feature in place, when using StatefulSet with maxUnavailable >1, the user is making a
+conscious choice that more than one pod going down at the same time during rolling update, would not
+cause issues with their Stateful Application which have per pod state and identity. Other Stateful
 Applications which cannot tolerate more than one pod going down, will resort to the current behavior of one pod at a time Rolling Updates.
 
 ### Goals
-StatefulSet RollingUpdate strategy will contain an additional parameter called `maxUnavailable` to 
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+StatefulSet RollingUpdate strategy will contain an additional parameter called `maxUnavailable` to
 control how many Pods will be brought down at a time, during Rolling Update.
 
 ### Non-Goals
-NA
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+None.
 
 ## Proposal
 
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
```

Comment:
We need specific details of the proposal, something like:
```
This document proposes to add new field to StatefulSet's RollingUpdate configuration 
settings (`.spec.updateStrategy.rollingUpdate`) called `maxUnavailable`, which will 
allow setting the maximum number of pods that can be unavailable during the update.
This new field can be an absolute, positive number (eg. `5`) or a percentage of desired 
pods (ex: `10%`). When the field is not specified, it will default to 1 to maintain 
previous behavior of rolling one pod at a time.
```